Batch serving and AWS deployment (#2251)
mlrun/mlrun: docs/serving/batch-serving.md

# Batch inference 

> This topic describes how to run a batch predict job along with drift analysis

Reviewer: Not sure if we need this sentence

## Overview

Batch inference or offline inference addresses the need to run machine learning model on large datasets. <br> 

> It is the process of generating predictions on a batch of observations. The batch jobs are typically generated on some recurring schedule (e.g. hourly, daily). <br>

Reviewer: Should mention early on that this is done by creating a job

These predictions are then stored in a database or a file and can be made available to developers or end users. Batch inference may sometimes take advantage <br>
of big data technologies such as Spark to generate predictions. This allows data scientists and machine learning engineers to take advantage of scalable compute <br>

> resources to generate many predictions at once. As opposed to real time inference real-time inference where the goal is usually to optimize the number of transactions per second <br>

Reviewer 2: ```suggestion
resources to generate many predictions at once. As opposed to real-time inference where the goal is usually to optimize the number of transactions per second <br>
```

> that the model can process,  With batch inference, the goal is usually tied to time constraints and the service-level agreement (SLA) for the job.  <br>

Reviewer 2: ```suggestion
that the model can process, with batch inference, the goal is usually tied to time constraints and the service-level agreement (SLA) for the job.  <br>
```

> in addition, the output of batch inference goes to a file or a table in a database vs sending a response back to an online application

Reviewer: Why so many line breaks?

## Auto-traininer function for prediction

> The [Auto-trainer](https://www.mlrun.org/marketplace/functions/master/auto_trainer/latest/example/) is a function in the mlrun marketplace that is used for both training and prediction. <br>

Reviewer 2: ```suggestion
The [Auto-trainer](https://www.mlrun.org/marketplace/functions/master/auto_trainer/latest/example/) is a function in the mlrun marketplace that is used for multiple steps in the model life-cycle including training, evaluation and prediction. <br>
```

This function saves you the need to write your own predict function and in addition it also runs a drift analysis by default. <br>

> The drift compares the prediction data with the sample/test dataset <br>

Reviewer 2: ```suggestion
The drift compares the prediction data with the sample/training dataset <br>
```

In order to run batch predict one should use the predict handler <br>

The function supports the following frameworks:
* XGBoost

> * sklearn

Reviewer 2: ```suggestion
* SciKit-Learn
```

* `result_set` - the name of the artifact with the result dataset.
* `param sample_set_statistics` (optional) - The statistics of the sample set logged along a model.
* `drift_threshold` (optional) - The threshold of which to mark drifts. Defaulted to 0.7
* `possible_drift_threshold` (optional) - The threshold of which to mark possible drifts. Defaulted to 0.5.

The output of the model is saved by default to the project artifact path but you can change it by changing the artifact_path parameters when you run the job. see [managing artifacts](https://docs.mlrun.org/en/latest/store/artifacts.html)

```python
import mlrun
```

```python
mlrun.get_or_create_project('batch-inference', context="./", user_project=True)
```

### Import auto-traininer function from the marketplace

```python
auto_trainer = mlrun.import_function("hub://auto_trainer")
```

### Input dataset

In this example we're using a csv file (stored on wassabi object store) as an input dataset for the prediction <br>

```python

> DATASET_URL = 'https://s3.wasabisys.com/iguazio/data/function-marketplace-data/xgb_trainer/classifier-data.csv'

Reviewer: Should use MLRun function to get sample data and not the full URL

      <td>-0.665400</td>
      <td>-1.157263</td>
      <td>0.346660</td>
      <td>-0.867286</td>
      <td>-1.903681</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>10000 rows × 6 columns</p>
</div>

Next, we drop the label column 

```python
sample = mlrun.get_dataitem(DATASET_URL).as_df().head().drop("labels", axis=1)
```

```python
sample = sample.values.tolist()
sample
```

    [[-0.2651151839702463,
      -1.9322600633270253,
      0.3039917127899224,
      -1.863833475610883,
      -1.045634803356314],
     [-3.1354794002707846,
      -2.835547936064964,
      1.3383808404873596,
      -1.385302821687631,
      -2.2764563589804325],
     [-1.519004884164945,
      -1.8075487992775987,
      0.6973035703821813,
      -1.1188598603515594,
      1.1048998961504315],
     [-0.6320867871545256,
      -0.3456588580241298,
      0.2443294437742147,
      -0.0460657045357599,
      0.4472804635971865],
     [-1.405882596126037,
      -1.7460450817666373,
      0.6536171315150338,
      -1.1109846976791669,
      -1.6754663728281285]]

## Run prediction

you can either use a trained model that is already registered in MLRun or just pass the model URI

```python

> model_path = 'store://artifacts/auto-trainer-admin/MyModel:cf9538d81a2a4ce9a2035ad3d8b90456'

Reviewer: Should reference a model from the samples and not something that is not runnable if I take the code as-is

    #csv table { border-collapse: collapse;}
    #csv table td { padding: 4px 8px; border: 1px solid silver;} </style>`;

  function csvToHtmlTable(str) {
    return '<div id="csv"><table><tr><td>' +  str.replace(/[\n\r]+$/g, '').replace(/[\n\r]+/g, '</td></tr><tr><td>')
      .replace(/,/g, '</td><td>') + '</td></tr></table></div>';
  }

  function reqListener () {
    if (el.title.endsWith(".csv")) {
      iframe.setAttribute("srcdoc", tblcss + csvToHtmlTable(this.responseText));
    } else {
      iframe.setAttribute("srcdoc", this.responseText);
    }
    console.log(this.responseText);
  }

  const oReq = new XMLHttpRequest();
  oReq.addEventListener("load", reqListener);
  oReq.open("GET", el.title);
  oReq.send();

  //iframe.src = el.title;
  const resultPane = document.querySelector(panelName + "-pane");
  if (resultPane.classList.contains("hidden")) {
    resultPane.classList.remove("hidden");
  }
}
function closePanel(el) {
  const panelName = "#" + el.getAttribute('paneName')
  const resultPane = document.querySelector(panelName + "-pane");
  if (!resultPane.classList.contains("hidden")) {
    resultPane.classList.add("hidden");
  }
}

</script>

> <div class="master-wrapper">

Reviewer: All of this renders in the markdown as text, need to fix

```python
predict_run = auto_trainer.run(
    params={
        "dataset": sample,
        "model": model_path,
        "drop_columns": [0, 2],
        "label_columns": "labels",
    },
    handler="predict",
    local=False,
    schedule='*/30 * * * *'
)
```

> 2022-08-11 06:01:09,183 [info] starting run auto-trainer-predict uid=9310360c64744c29991d1145d04d55e5 DB=http://mlrun-api:8080
> 2022-08-11 06:01:09,456 [info] task scheduled, {'schedule': '*/30 * * * *', 'project': 'batch-inference-admin', 'name': 'auto-trainer-predict'}

## Viewing the drift result

By default, the auto_trainer functions analyze drift by comparing the input and the sample dataset statistics. <br> 
it triggers a drift analysis job which generates 3 output files that can be viewed under the artifacts tab for the prediction job
* drift_table_plot - shows expected vs actual statistics for every feature as a plot.
* drift_results - shows the drift result per feature as a JSON format so you can use it to process this data and apply your logic on top of it.

> * dataset_statistics - shows the statistics of the input dataset.

Reviewer: Should also show how to present the artifact here

Reviewer 2: should be done with something like:

```
predict_run_object.artifact('drift_table_plot').show()
```

In the next section Drift table plot example
